File: 435-non-overlapping-intervals.java
class Solution {
    public int eraseOverlapIntervals(int[][] intervals) {
       if(intervals.length == 0) { return 0; }
        
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[1], b[1]));
        
        int end = intervals[0][1];
        int count = 1;
        for(int i = 1; i < intervals.length; i++) {
            int[] interval = intervals[i];
            if(interval[0] >= end) {
                count++;
                end = interval[1];
            }
        }
        
        return intervals.length - count;
    }
}

File: 435-non-overlapping-intervals.py
To minimize removals, I maximize the number of non-overlapping intervals.
A greedy strategy works: sort by end time and always keep the interval that ends earliest.

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x : x[1])
        prev_end = intervals[0][1]
        removed = 0
        for start, end in intervals[1:]:
            if start < prev_end:
                removed += 1
            else:
                prev_end = end
        
        return removed

